virtualizarr/manifests/manifest.py

```diff
@@ -1,65 +1,118 @@
+            Required if any (likely kerchunk-generated) paths are relative in order to turn them into absolute paths (which virtualizarr requires).
+        """
 
-    Stored in the form `{"path": "s3://bucket/foo.nc", "offset": 100, "length": 100}`.
-    """
+        # note: we can't just use `__init__` or a dataclass' `__post_init__` because we need `fs_root` to be an optional kwarg
 
-    path: str  # TODO stricter typing/validation of possible local / remote paths?
-    offset: int
-    length: int
+        path = validate_and_normalize_path_to_uri(path, fs_root=fs_root)
 
-    @classmethod
-    def from_kerchunk(
-        cls, path_and_byte_range_info: tuple[str] | tuple[str, int, int]
-    ) -> "ChunkEntry":
-        from upath import UPath
+        if isinstance(offset, np.integer):
+            _offset = int(offset)
+        elif isinstance(offset, int):
+            _offset = offset
+        else:
+            raise TypeError(
+                f"chunk entry byte offset must of type int, but got type {type(offset)}"
+            )
+        if _offset < 0:
+            raise ValueError(
+                f"chunk entry byte offset must be a positive integer, but got offset={_offset}"
+            )
 
-        if len(path_and_byte_range_info) == 1:
-            path = path_and_byte_range_info[0]
-            offset = 0
-            length = UPath(path).stat().st_size
+        if isinstance(length, np.integer):
+            _length = int(length)
+        elif isinstance(length, int):
+            _length = length
         else:
-            path, offset, length = path_and_byte_range_info
+            raise TypeError(
+                f"chunk entry byte offset must of type int, but got type {type(length)}"
+            )
+        if _length < 0:
+            raise ValueError(
+                f"chunk entry byte offset must be a positive integer, but got offset={_length}"
+            )
+
         return ChunkEntry(path=path, offset=offset, length=length)
 
-    def to_kerchunk(self) -> tuple[str, int, int]:
-        """Write out in the format that kerchunk uses for chunk entries."""
-        return (self.path, self.offset, self.length)
 
-    def dict(self) -> ChunkDictEntry:
-        return ChunkDictEntry(
-            path=self.path,
-            offset=self.offset,
-            length=self.length,
-        )
+def validate_and_normalize_path_to_uri(path: str, fs_root: str | None = None) -> str:
+    """
+    Makes all paths into fully-qualified absolute URIs, or raises
+
+    See https://en.wikipedia.org/wiki/File_URI_scheme
+
+    Parameters
+    ----------
+    fs_root
+        The root of the filesystem on which these references were generated.
+        Required if any (likely kerchunk-generated) paths are relative in order to turn them into absolute paths (which virtualizarr requires).
+    """
+    if not any(path.startswith(prefix) for prefix in VALID_URI_PREFIXES) and path != "":
```

Comment:
I've seen people use urlparse (https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse). Not sure if that would be easier to read or not.

Reply:
Actually handling all these cases is such a PITA that maybe I should just use [`cloudpathlib.AnyPath`](https://cloudpathlib.drivendata.org/stable/anypath-polymorphism/)

Reply:
Using cloudpathlib did make this very easy, at the cost of adding a new dependency. Luckily for this use case (string validation) it's only a pure-python standalone dependency.